Add new trigger link mapping file for updated backend

snwebb opened the pull request:
#### PR description:

Add new trigger link mapping file for the updated backend layout of 14 Stage 1 FPGAs per 120º sector.

snwebb committed:
Create hgcal_trigger_link_mapping_v2.json

jbsauvan commented:
If possible it would be good to now use the number of input links (here, 120) in the file name, since now there will be several different scenarios

snwebb committed:
Merge branch 'PFCal-dev/hgc-tpg-devel-CMSSW_12_0_0_pre3' into snwebb-update-json-mapping

snwebb committed:
rename new mapping file

snwebb committed:
add new function for finding scintillator ieta iphi

snwebb committed:
fix bugs

snwebb commented:
Hi @jbsauvan, the pull request is still in progress, but are the implemented changes as you might have expected for the scintillator?

snwebb committed:
add new-version mapping files for 120 and 72 links

snwebb committed:
remove magic numbers and tidy up

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -112,6 +118,8 @@ class HGCalTriggerGeometryV9Imp3 : public HGCalTriggerGeometryBase {
   HGCalGeomRotation::WaferCentring getWaferCentring(unsigned layer, int subdet) const;
   void etaphiMappingFromSector0(int& ieta, int& iphi, unsigned sector) const;
   unsigned etaphiMappingToSector0(int& ieta, int& iphi) const;
```
It seems that `etaphiMappingToSector0` is not used anymore. Is `tcEtaphiMappingToSector0` replacing it?

snwebb replied:
Yes, I left the function in, in case it was useful for anyone in the future, given it would still work. But we don't need it now at present.

snwebb replied:
`tcEtaphiMappingToSector0` and `getScintillatoriEtaiPhi` together replace it

jbsauvan replied:
Ok. In that case it might be better to remove the old function

snwebb replied:
Ok, done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -112,6 +118,8 @@ class HGCalTriggerGeometryV9Imp3 : public HGCalTriggerGeometryBase {
   HGCalGeomRotation::WaferCentring getWaferCentring(unsigned layer, int subdet) const;
   void etaphiMappingFromSector0(int& ieta, int& iphi, unsigned sector) const;
```
`etaphiMappingFromSector0` is used to get trigger cells from modules but it doesn't look like the inverse of `tcEtaphiMappingToSector0` which is used to get modules from trigger cells. Shouldn't it be changed as well?

snwebb replied:
etaphiMappingFromSector0 is used to rotate the scintillator module, not the trigger cells. `hSc_num_panels_per_sector_` should be 12 now though I think, so I've updated it.

I don't think we need an inverse of `tcEtaphiMappingToSector0` since we get use the `getTriggerCellsFromModule` once we've rotated the module to sector 0.

jbsauvan replied:
Ok, thanks

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -68,6 +68,12 @@ class HGCalTriggerGeometryV9Imp3 : public HGCalTriggerGeometryBase {
   unsigned hSc_triggercell_size_ = 2;
   unsigned hSc_module_size_ = 12;  // in TC units (144 TC / panel = 36 e-links)
   unsigned hSc_wafers_per_module_ = 3;
+  unsigned hSc_modules_per_sector_ = 4;
+  unsigned hSc_front_layers_split_ = 12;
+  unsigned hSc_back_layers_split_ = 8;
+  unsigned hSc_layer_for_split_ = 40;
+  int hSc_tc_layer0_min_ = 24;
+  const int ntc_per_wafer = 48;
```
It seems that these variables are not modified. If that's the case they should be defined as `static constexpr`

snwebb replied:
Done, apart from the first 3 which can be set in the config file

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -68,6 +68,12 @@ class HGCalTriggerGeometryV9Imp3 : public HGCalTriggerGeometryBase {
   unsigned hSc_triggercell_size_ = 2;
   unsigned hSc_module_size_ = 12;  // in TC units (144 TC / panel = 36 e-links)
   unsigned hSc_wafers_per_module_ = 3;
+  unsigned hSc_modules_per_sector_ = 4;
+  unsigned hSc_front_layers_split_ = 12;
+  unsigned hSc_back_layers_split_ = 8;
+  unsigned hSc_layer_for_split_ = 40;
+  int hSc_tc_layer0_min_ = 24;
+  const int ntc_per_wafer = 48;
```
Missing trailing underscore for `ntc_per_wafer` (as it is a class member)

snwebb replied:
Done

jbsauvan commented:
> Hi @jbsauvan, the pull request is still in progress, but are the implemented changes as you might have expected for the scintillator?

Quickly looking at https://github.com/snwebb/hgcal-linkmapping/blob/master/fluctuation.py#L119-L161, it seems indeed to correspond to what is needed.
But I suppose the function `etaphiMappingFromSector0` would also need to be updated (see my previous comments) in order to correctly get trigger cells from module, or is the existing implementation still valid?

snwebb committed:
changes after comments

snwebb commented:
thanks @jbsauvan, the comments have been implemented

jbsauvan commented:
Thanks @snwebb 
It seems that the geometry tester `L1THGCal/test/testHGCalL1TGeometryV11Imp3_cfg.py` fails in the cell <-> module in the scintillator [1]. Which is a bit strange since the TC <-> module and the cell <-> TC are passing.
Would it be possible to have a look?

[1]
```
Checking cell -> trigger cell -> cell consistency
Checking trigger cell -> module -> trigger cell consistency
Checking cell -> module -> cell consistency
Error:
 Cell 2787921184( HGCScintillatorDetId::EE:HE= 0:1 trigger= 0 type= 1 SiPM= 0 layer= 22 ring= -38:-38 phi= 288:288)
 has not been found in
 module HGCalTriggerModuleDetId::HFNose:EE:HSil:HScin= 0:0:0:1 type= 1 z= -1 layer= 22 sector= 1 module(u,v)= (1,9)
 Available cells are:
2787905274 2787905273 2787905272 2787905271 2787905270 2787905269 2787905268 2787905267 2787905266 2787905265 2787906300 2787906299 2787906811 2787906812 2787904763 2787904764 2787903218 2787903217 2787903729 2787903730 2787904761 2787904762 2787904759 2787904760 2787904757 2787904758 2787904755 2787904756 2787902705 2787907323 2787906289 2787902707 2787902706 2787907324 2787906291 2787902709 2787906290 2787902708 2787903221 2787906803 2787904244 2787907826 2787905782 2787902200 2787903220 2787906802 2787907836 2787904243 2787907825 2787905781 2787902199 2787905788 2787904245 2787907827 2787905783 2787902201 2787904246 2787907828 2787905784 2787902202 2787904247 2787907829 2787905785 2787902203 2787904252 2787907834 2787903735 2787907317 2787903219 2787906801 2787907835 2787904250 2787907832 2787903733 2787907315 2787904251 2787907833 2787903734 2787907316 2787904248 2787907830 2787905786 2787902204 2787904241 2787903731 2787907313 2787905276 2787904249 2787907831 2787904242 2787903732 2787907314 2787907320 2787903738 2787907319 2787903737 2787906807 2787903225 2787906808 2787903226 2787907318 2787903736 2787906805 2787903223 2787906806 2787903224 2787906804 2787903222 2787906298 2787902716 2787904753 2787906297 2787902715 2787905777 2787902195 2787905778 2787902196 2787906292 2787902710 2787905779 2787902197 2787905780 2787902198 2787905787 2787906294 2787902712 2787906293 2787902711 2787906296 2787902714 2787906295 2787902713 2787902193 2787902194 2787907322 2787903740 2787907321 2787903739 2787906809 2787903227 2787906810 2787903228 2787905275 2787904754
%MSG-w HGCalTriggerGeometryTester:   HGCalTriggerGeomTesterV9Imp3:hgcaltriggergeomtester@streamBeginRun  20-Aug-2021 09:26:09 CEST Run: 1 Stream: 0
Problem with the trigger geometry detected. Only the basic cells tree will be filled

%MSG
%MSG-w HGCalTriggerGeometryTester:   HGCalTriggerGeomTesterV9Imp3:hgcaltriggergeomtester@streamBeginRun  20-Aug-2021 09:26:09 CEST Run: 1 Stream: 0
HGCalTriggerGeometry: Found inconsistency in cell <-> module mapping
```

snwebb committed:
some fixes for scintillator but still get error

snwebb committed:
fix bug in tester

snwebb committed:
scintillator geometry fixed, plus fix bug in stage1 links to stage2

snwebb commented:
Hi @jbsauvan, the tester now runs without errors. I corrected the scintillator geometry and also found a bug in the stage 2 to stage 1 links, which is now also corrected.
